cmd/openstack-populator/openstack-populator.go

```diff
@@ -11,136 +11,174 @@ import (
+	flag.StringVar(&config.ownerUID, "owner-uid", "", "Owner UID (usually PVC UID)")
+	flag.Int64Var(&config.pvcSize, "pvc-size", 0, "Size of pvc (in bytes)")
 	flag.Parse()
 
-	populate(volumePath, identityEndpoint, secretName, imageID)
+	return config
 }
 
-func populate(fileName, identityEndpoint, secretName, imageID string) {
+func initializeMetrics() {
 	http.Handle("/metrics", promhttp.Handler())
 	go http.ListenAndServe(":2112", nil)
-	progressGague := prometheus.NewGaugeVec(
-		prometheus.GaugeOpts{
-			Subsystem: "volume_populators",
-			Name:      "openstack_volume_populator",
-			Help:      "Amount of data transferred",
-		},
-		[]string{"image_id"},
-	)
+}
 
-	if err := prometheus.Register(progressGague); err != nil {
-		klog.Error("Prometheus progress counter not registered:", err)
-	} else {
-		klog.Info("Prometheus progress counter registered.")
-	}
+func populate(config *AppConfig) {
+	client := createClient(config)
+	downloadAndSaveImage(client, config)
+}
 
+func createClient(config *AppConfig) *libclient.Client {
 	options := readOptions()
-
 	client := &libclient.Client{
-		URL:     identityEndpoint,
+		URL:     config.identityEndpoint,
 		Options: options,
 	}
-
 	err := client.Connect()
 	if err != nil {
 		klog.Fatal(err)
 	}
+	return client
+}
 
-	klog.Info("Downloading the image: ", imageID)
-	imageReader, err := client.DownloadImage(imageID)
+func downloadAndSaveImage(client *libclient.Client, config *AppConfig) {
+	klog.Info("Downloading the image: ", config.imageID)
+	imageReader, err := client.DownloadImage(config.imageID)
 	if err != nil {
 		klog.Fatal(err)
 	}
 	defer imageReader.Close()
 
-	if err != nil {
-		klog.Fatal(err)
+	file := openFile(config.volumePath)
+	defer file.Close()
+
+	progressVec := createProgressCounter()
+	writeData(imageReader, file, config, progressVec)
+}
+
+func createProgressCounter() *prometheus.CounterVec {
+	progressVec := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "openstack_populator_progress",
+			Help: "Progress of volume population",
+		},
+		[]string{"ownerUID"},
+	)
+	if err := prometheus.Register(progressVec); err != nil {
+		klog.Error("Prometheus progress counter not registered:", err)
+	} else {
+		klog.Info("Prometheus progress counter registered.")
 	}
+	return progressVec
+}
+
+func openFile(volumePath string) *os.File {
 	flags := os.O_RDWR
-	if strings.HasSuffix(fileName, "disk.img") {
+	if strings.HasSuffix(volumePath, "disk.img") {
 		flags |= os.O_CREATE
 	}
-
-	klog.Info("Saving the image to: ", fileName)
-	file, err := os.OpenFile(fileName, flags, 0650)
+	file, err := os.OpenFile(volumePath, flags, 0650)
 	if err != nil {
 		klog.Fatal(err)
 	}
-	defer file.Close()
+	return file
+}
 
-	err = writeData(imageReader, file, imageID, progressGague)
-	if err != nil {
+func writeData(reader io.ReadCloser, file *os.File, config *AppConfig, progress *prometheus.CounterVec) {
```

Comment:
shouldn't we defer file.Close() here?

Reply:
it's done in the caller `downloadAndSaveImage`

Reply:
ack